core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java

```diff
@@ -85,21 +88,24 @@
     private final DiscoverySettings discoverySettings;
     private final ClusterName clusterName;
     private final PendingClusterStatesQueue pendingStatesQueue;
+    private final CustomPrototypeRegistry registry;
 
     public PublishClusterStateAction(
             Settings settings,
             TransportService transportService,
             Supplier<ClusterState> clusterStateSupplier,
             NewPendingClusterStateListener listener,
             DiscoverySettings discoverySettings,
-            ClusterName clusterName) {
+            ClusterName clusterName,
+            CustomPrototypeRegistry registry) {
```

Comment:
The main reason I needed to provide `CustomPrototypeRegistry` here is because in the `handleIncomingClusterStateRequest(...)` method (line 380-385) it creates a new `StreamInput` instance and have to wrap that with a `NamedWriteableAwareStreamInput` instance.